subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishIssuesIntegTest.groovy

```diff
@@ -62,4 +62,106 @@ public class IvyPublishIssuesIntegTest extends AbstractIvyPublishIntegTest {
+    void "doesn't publish if main artifact is missing"() {
+        settingsFile << 'rootProject.name = "test"'
+        buildFile << """
+            apply plugin: "java-library"
+            apply plugin: "ivy-publish"
+
+            group = "org.gradle"
+            version = "1.0"
+
+            jar {
+                enabled = Boolean.valueOf(project.getProperty("jarEnabled"))
+            }
+
+            publishing {
+                repositories {
+                    ivy { url "\${buildDir}/repo" }
+                }
+                publications {
+                    ivy(IvyPublication) {
+                        from components.java
+                    }
+                }
+            }
+        """
+        file("src/main/java/hello/Hello.java") << """package hello;
+            public class Hello {}
+        """
+
+        when:
+        succeeds "publish", "-PjarEnabled=true"
+
+        then:
+        file("build/repo/org.gradle/test/1.0/test-1.0.jar").exists()
+        file("build/repo/org.gradle/test/1.0/test-1.0.jar.sha1").exists()
+        file("build/repo/org.gradle/test/1.0/test-1.0.jar.sha256").exists()
+        file("build/repo/org.gradle/test/1.0/test-1.0.jar.sha512").exists()
+
+        when:
+        fails "publish", "-PjarEnabled=false"
+
+        then:
+        skipped(":jar")
+        failure.assertHasCause("Artifact test-1.0.jar wasn't produced by this build.")
+    }
+
+    @ToBeFixedForInstantExecution
+    @Issue("https://github.com/gradle/gradle/issues/5136")
+    void "doesn't publish stale files"() {
+        settingsFile << 'rootProject.name = "test"'
+        buildFile << """
+            apply plugin: "java-library"
+            apply plugin: "ivy-publish"
+
+            group = "org.gradle"
+            version = "1.0"
+
+            java {
+                withJavadocJar()
+            }
+
+            javadocJar {
+                enabled = Boolean.valueOf(project.getProperty("javadocEnabled"))
+            }
+
+            publishing {
+                repositories {
+                    ivy { url "\${buildDir}/repo" }
+                }
+                publications {
+                    ivy(IvyPublication) {
+                        from components.java
+                    }
+                }
+            }
+        """
+        file("src/main/java/hello/Hello.java") << """package hello;
+            public class Hello {}
+        """
+
+        when:
+        succeeds "publish", "-PjavadocEnabled=true"
+
+        then:
+        file("build/repo/org.gradle/test/1.0/test-1.0-javadoc.jar").exists()
+        file("build/repo/org.gradle/test/1.0/test-1.0-javadoc.jar.sha1").exists()
+        file("build/repo/org.gradle/test/1.0/test-1.0-javadoc.jar.sha256").exists()
+        file("build/repo/org.gradle/test/1.0/test-1.0-javadoc.jar.sha512").exists()
+
+        when:
+        file("build/repo").deleteDir()
+        succeeds "publish", "-PjavadocEnabled=false"
```

Comment:
:o: In this example, this leads to illegal module metadata that is published. The metadata still includes the reference to "test-1.0-javadoc.jar" in the "javadocElements" variant.

I am not sure if this was already broken before or if it would have failed, maybe this can be easily fixed by validating that the files exist before writing the module metadata?

If it is more involved, we can also do a follow up PR for 6.2 as well.

Reply:
fixed

Reply:
Great! That fix was simple than expected. :)